libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizer.cs

```diff
@@ -0,0 +1,217 @@
+            if (compositeEntities != null && compositeEntities.Any())
+            {
+                compositeEntityTypes = new HashSet<string>(compositeEntities.Select(ce => ce.ParentType));
+                entities = compositeEntities.Aggregate(entities, (current, compositeEntity) => PopulateCompositeEntity(compositeEntity, current, entitiesAndMetadata, verbose));
+            }
+
+            foreach (var entity in entities)
+            {
+                // we'll address composite entities separately
+                if (compositeEntityTypes.Contains(entity.Type))
+                    continue;
+
+                AddProperty(entitiesAndMetadata, GetNormalizedEntityType(entity), GetEntityValue(entity));
+
+                if (verbose)
+                {
+                    AddProperty((JObject) entitiesAndMetadata["$instance"], GetNormalizedEntityType(entity), GetEntityMetadata(entity));
+                }
+            }
+
+            return entitiesAndMetadata;
+        }
+
+        private static JToken GetEntityValue(EntityRecommendation entity)
+        {
+            if (entity.Resolution == null)
+                return entity.Entity;
+
+            if (entity.Type.StartsWith("builtin.datetimeV2."))
+            {
+                return new JValue(entity.Resolution?.Values != null && entity.Resolution.Values.Count > 0
+                            ? ((IDictionary<string, object>)((IList<object>)entity.Resolution.Values.First()).First())["timex"]
+                            : entity.Resolution);
+            }
+            
+            if (entity.Type.StartsWith("builtin.number"))
+            {
+                var value = (string) entity.Resolution.Values.First();
+                return long.TryParse(value, out var longVal) ?
+                            new JValue(longVal) :
+                            new JValue(double.Parse(value));
+            }
+
+            return entity.Resolution.Count > 1 ? 
+                JObject.FromObject(entity.Resolution) : 
+                entity.Resolution.ContainsKey("value") ?
+                    (JToken) JObject.FromObject(entity.Resolution["value"]) :
+                    JArray.FromObject(entity.Resolution["values"]);
+        }
+
+        private static JObject GetEntityMetadata(EntityRecommendation entity)
+        {
+            return JObject.FromObject(new
+            {
+                startIndex = entity.StartIndex,
+                endIndex = entity.EndIndex,
+                text = entity.Entity,
+                score = entity.Score
+            });
+        }
+
+        private static string GetNormalizedEntityType(EntityRecommendation entity)
+        {
+            return Regex.Replace(entity.Type, "\\.", "_");
+        }
+
+        private static IList<EntityRecommendation> PopulateCompositeEntity(CompositeEntity compositeEntity, IList<EntityRecommendation> entities, JObject entitiesAndMetadata, bool verbose)
```

Comment:
Consider breaking up into sub-mehods.

Reply:
I made the method shorter, I think it would make it harder to follow if I break it any further.